File: services/inventory/internal/adapter/kafka/producer.go
package kafka

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
	"go.opentelemetry.io/contrib/bridges/otelslog"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"
)

const instrumentationName = "enterprise-oss-lab/sample-ec-service/inventry/internal/adapter/kafka"

// KafkaHeaderCarrier adapts a slice of confluent-kafka Headers to the
// propagation.TextMapCarrier interface so trace context can be injected into
// (producer) and extracted from (consumer) Kafka message headers manually.
// It holds a pointer to the header slice so Set can grow it in place.
type KafkaHeaderCarrier struct {
	headers *[]kafka.Header
}

// NewKafkaHeaderCarrier wraps the given header slice.
func NewKafkaHeaderCarrier(headers *[]kafka.Header) KafkaHeaderCarrier {
	return KafkaHeaderCarrier{headers: headers}
}

// Get returns the value for the given header key, or "" if absent.
func (c KafkaHeaderCarrier) Get(key string) string {
	for _, h := range *c.headers {
		if h.Key == key {
			return string(h.Value)
		}
	}
	return ""
}

// Set upserts a header key/value pair.
func (c KafkaHeaderCarrier) Set(key, value string) {
	for i := range *c.headers {
		if (*c.headers)[i].Key == key {
			(*c.headers)[i].Value = []byte(value)
			return
		}
	}
	*c.headers = append(*c.headers, kafka.Header{Key: key, Value: []byte(value)})
}

// Keys lists all header keys.
func (c KafkaHeaderCarrier) Keys() []string {
	keys := make([]string, 0, len(*c.headers))
	for _, h := range *c.headers {
		keys = append(keys, h.Key)
	}
	return keys
}

type Producer struct {
	producer *kafka.Producer
	topic    string
	tracer   trace.Tracer
	logger   *slog.Logger
}

func NewProducer(brokers []string, topic string) (*Producer, error) {
	p, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers": strings.Join(brokers, ","),
		"acks":              "all",
	})
	if err != nil {
		return nil, fmt.Errorf("new producer: %w", err)
	}
	return &Producer{
		producer: p,
		topic:    topic,
		tracer:   otel.Tracer(instrumentationName),
		logger:   otelslog.NewLogger(instrumentationName),
	}, nil
}

func (p *Producer) Publish(ctx context.Context, result ReservationResult) error {
	ctx, span := p.tracer.Start(ctx, "produce "+p.topic, trace.WithSpanKind(trace.SpanKindProducer))
	defer span.End()

	span.SetAttributes(
		attribute.String("messaging.system", "kafka"),
		attribute.String("messaging.destination.name", p.topic),
		attribute.String("messaging.operation", "publish"),
		attribute.String("messaging.kafka.message.key", result.CorrelationID),
	)

	payload, err := json.Marshal(result)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return fmt.Errorf("marshal result: %w", err)
	}

	msg := &kafka.Message{
		TopicPartition: kafka.TopicPartition{Topic: &p.topic, Partition: kafka.PartitionAny},
		Key:            []byte(result.CorrelationID),
		Value:          payload,
	}

	// Inject the current trace context as W3C traceparent into the message headers.
	otel.GetTextMapPropagator().Inject(ctx, NewKafkaHeaderCarrier(&msg.Headers))

	deliveryCh := make(chan kafka.Event, 1)
	if err := p.producer.Produce(msg, deliveryCh); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return fmt.Errorf("produce: %w", err)
	}

	select {
	case e := <-deliveryCh:
		m := e.(*kafka.Message)
		if m.TopicPartition.Error != nil {
			span.RecordError(m.TopicPartition.Error)
			span.SetStatus(codes.Error, m.TopicPartition.Error.Error())
			return fmt.Errorf("delivery: %w", m.TopicPartition.Error)
		}
	case <-ctx.Done():
		span.RecordError(ctx.Err())
		span.SetStatus(codes.Error, ctx.Err().Error())
		return ctx.Err()
	}

	p.logger.InfoContext(ctx, "published reservation result",
		"topic", p.topic,
		"correlation_id", result.CorrelationID,
		"success", result.Success,
	)
	return nil
}

func (p *Producer) Close() error {
	p.producer.Flush(5000)
	p.producer.Close()
	return nil
}

File: services/order/internal/adapter/kafka/producer.py
import asyncio
import logging

from confluent_kafka import Producer
from opentelemetry import propagate, trace

from internal.adapter.kafka.message import ReservationRequest
from internal.usecase.order import KafkaProducerPort

logger = logging.getLogger(__name__)
tracer = trace.get_tracer(__name__)


class KafkaReservationProducer(KafkaProducerPort):
    def __init__(self, brokers: str, topic: str) -> None:
        self._producer = Producer({"bootstrap.servers": brokers, "acks": "1"})
        self._topic = topic

    async def publish_reservation_request(
        self,
        correlation_id: str,
        inventory_id: int,
        quantity: int,
    ) -> None:
        msg = ReservationRequest(
            correlation_id=correlation_id,
            inventory_id=inventory_id,
            quantity=quantity,
        )
        payload = msg.model_dump_json().encode()

        def delivery_report(err, _msg):
            if err:
                logger.error("kafka delivery failed: %s", err)

        with tracer.start_as_current_span(
            f"{self._topic} publish",
            kind=trace.SpanKind.PRODUCER,
            attributes={
                "messaging.system": "kafka",
                "messaging.destination.name": self._topic,
                "messaging.operation": "publish",
            },
        ):
            # trace context を W3C traceparent として Kafka ヘッダに注入 (手動伝播)
            carrier: dict[str, str] = {}
            propagate.inject(carrier)
            headers = [(k, v.encode()) for k, v in carrier.items()]

            # produce() はローカルキューに積むだけの非ブロッキング呼び出し。
            # rdkafka のバックグラウンドスレッドが実際の送信を行う。
            self._producer.produce(
                self._topic,
                key=correlation_id.encode(),
                value=payload,
                headers=headers,
                callback=delivery_report,
            )

        # 配信レポートのコールバックを捌く。timeout=0 なので非ブロッキングで、
        # イベントループを止めない (以前の flush(timeout=5) はループ全体を最大5秒ブロックしていた)。
        self._producer.poll(0)

    async def run(self) -> None:
        """配信レポートを定期的に処理し続けるバックグラウンドループ。

        publish 側の poll(0) だけでも捌けるが、無通信区間でも内部キューを
        確実にドレインするために常駐させる。イベントループはブロックしない。
        """
        while True:
            self._producer.poll(0)
            await asyncio.sleep(0.1)

    def close(self) -> None:
        # 終了時のみ、未送信メッセージを送り切るために同期 flush する。
        self._producer.flush(timeout=10)
